src/aws-secrets-controller/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+    let sdk_config = load_sdk_config(region).await;
+    let sm_config = SecretsManagerConfigBuilder::from(&sdk_config).build();
+    Client::from_conf(sm_config)
+}
+
+impl AwsSecretsController {
+    pub async fn new(
+        environment_id: EnvironmentId,
+        default_tags: BTreeMap<String, String>,
+    ) -> Self {
+        AwsSecretsController {
+            client: load_secrets_manager_client(environment_id.cloud_provider_region().to_owned())
+                .await,
+            secret_name_prefix: format!("/user-managed/{}/", environment_id),
+            environment_id,
+            default_tags,
+        }
+    }
+    fn tags(&self) -> Vec<Tag> {
+        self.default_tags
+            .iter()
+            .map(|(key, value)| Tag::builder().key(key).value(value).build())
+            .collect()
+    }
+
+    fn kms_key_alias(&self) -> String {
+        // Do not change this, as it must match the code in the region-controller.
+        format!("alias/customer_key_{}", &self.environment_id)
+    }
+
+    fn secret_name(&self, id: GlobalId) -> String {
+        format!("{}{}", self.secret_name_prefix, id)
+    }
+
+    fn id_from_secret_name(&self, name: &str) -> Option<GlobalId> {
+        name.strip_prefix(&self.secret_name_prefix)
+            .and_then(|id| id.parse().ok())
+    }
+
+    // TODO [Alex Hunt] Remove after all customers have been migrated.
+    pub async fn migrate_from<C>(&self, other_secrets_controller: &C) -> Result<(), anyhow::Error>
+    where
+        C: SecretsController,
+    {
+        let other_secrets: BTreeSet<GlobalId> =
+            other_secrets_controller.list().await?.into_iter().collect();
+
+        let aws_secrets: BTreeSet<GlobalId> = self.list().await?.into_iter().collect();
+
+        let other_secrets_reader = other_secrets_controller.reader();
+        let secrets_to_migrate = other_secrets.difference(&aws_secrets);
+        for secret in secrets_to_migrate {
+            let data = other_secrets_reader.read(secret.to_owned()).await?;
+            self.ensure(secret.to_owned(), &data).await?;
+        }
+        for secret in other_secrets {
+            other_secrets_controller.delete(secret).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SecretsController for AwsSecretsController {
+    async fn ensure(&self, id: GlobalId, contents: &[u8]) -> Result<(), anyhow::Error> {
+        match self
+            .client
+            .create_secret()
+            .name(self.secret_name(id))
+            .kms_key_id(self.kms_key_alias())
+            .secret_binary(Blob::new(contents))
+            .set_tags(Some(self.tags()))
+            .send()
+            .await
+        {
+            Ok(_) => {}
+            Err(SdkError::ServiceError(e)) if e.err().is_resource_exists_exception() => {
+                self.client
+                    .put_secret_value()
+                    .secret_id(self.secret_name(id))
+                    .secret_binary(Blob::new(contents))
```

Comment:
Do we need to set / update tags here, or is it a safe assumption the tags set at creation were correct?